From 60dfc0a48a3a3fead7ebccaf50d62ea03fe64403 Mon Sep 17 00:00:00 2001
From: Pekka Vuorela <pekka.vuorela@jolla.com>
Date: Wed, 23 Jan 2019 16:25:10 +0200
Subject: [PATCH] Use systemd activation for dbus. Contributes to JB#24120

Add systemd unit file and adjust building and packaging accordingly.
Add SystemdService to dbus service.
---
 Makefile.am                       |  4 ++++
 configure.ac                      | 12 +++++++++++-
 src/dbus-org.pacrunner.service.in |  8 ++++++++
 src/org.pacrunner.service.in      |  1 +
 4 files changed, 24 insertions(+), 1 deletion(-)
 create mode 100644 src/dbus-org.pacrunner.service.in

diff --git a/Makefile.am b/Makefile.am
index 2b99595..e1121fd 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -9,6 +9,10 @@ dbusconf_DATA = src/pacrunner.conf
 dbusdatadir = @DBUS_DATADIR@
 
 dbusdata_DATA = src/org.pacrunner.service
+
+systemdunitdir = @SYSTEMD_UNITDIR@
+
+systemdunit_DATA = src/dbus-org.pacrunner.service
 endif
 
 gdbus_sources = gdbus/gdbus.h gdbus/mainloop.c gdbus/watch.c \
diff --git a/configure.ac b/configure.ac
index 717dc00..d5eb8c4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -122,6 +122,16 @@ else
 fi
 AC_SUBST(DBUS_DATADIR)
 
+AC_ARG_WITH(systemdunitdir, AC_HELP_STRING([--with-systemdunitdir=PATH],
+	[path to systemd unit directory]), [path_systemdunit=${withval}],
+		[path_systemdunit="`$PKG_CONFIG systemd --variable=systemdsystemunitdir`"])
+if (test -z "${path_systemdunit}"); then
+	SYSTEMD_UNITDIR="/lib/systemd/system"
+else
+	SYSTEMD_UNITDIR="${path_systemdunit}"
+fi
+AC_SUBST(SYSTEMD_UNITDIR)
+
 AC_CHECK_LIB(pthread, pthread_create,
 	[PTHREAD_LIBS="-lpthread"],
 	[AC_MSG_ERROR([Pthread support is missing])])
@@ -150,4 +160,4 @@ AC_ARG_ENABLE(datafiles, AC_HELP_STRING([--disable-datafiles],
 					[enable_datafiles=${enableval}])
 AM_CONDITIONAL(DATAFILES, test "${enable_datafiles}" != "no")
 
-AC_OUTPUT(Makefile src/org.pacrunner.service libproxy/libproxy-1.0.pc)
+AC_OUTPUT(Makefile src/org.pacrunner.service src/dbus-org.pacrunner.service libproxy/libproxy-1.0.pc)
diff --git a/src/dbus-org.pacrunner.service.in b/src/dbus-org.pacrunner.service.in
new file mode 100644
index 0000000..9c87f89
--- /dev/null
+++ b/src/dbus-org.pacrunner.service.in
@@ -0,0 +1,13 @@
+[Unit]
+Description=Pacrunner service
+
+[Service]
+Type=dbus
+BusName=org.pacrunner
+ExecStart=@prefix@/sbin/pacrunner -n
+# Sandboxing
+CapabilityBoundingSet=CAP_NET_RAW
+PrivateDevices=true
+PrivateTmp=yes
+ProtectHome=yes
+ProtectSystem=full
diff --git a/src/org.pacrunner.service.in b/src/org.pacrunner.service.in
index eafd39d..01f9ef4 100644
--- a/src/org.pacrunner.service.in
+++ b/src/org.pacrunner.service.in
@@ -1,4 +1,5 @@
 [D-BUS Service]
 Name=org.pacrunner
-Exec=@prefix@/sbin/pacrunner
+Exec=/bin/false
 User=root
+SystemdService=dbus-org.pacrunner.service
-- 
2.20.1

